File: benchmarks/rust/rbtree_del.rs
#[derive(Clone, Copy)]
enum Color {
    Red,
    Black,
}

#[derive(Clone)]
enum Tree {
    Leaf,
    Node(Color, Box<Tree>, i64, bool, Box<Tree>),
}

#[derive(Clone)]
struct Del(Tree, bool);

#[derive(Clone)]
struct Delmin(Del, i64, bool);

const LIMIT: i64 = 100_000;

fn count_true(tree: &Tree) -> i64 {
    match tree {
        Tree::Leaf => 0,
        Tree::Node(_, left, _, value, right) => count_true(left) + i64::from(*value) + count_true(right),
    }
}

fn is_red(tree: &Tree) -> bool {
    matches!(tree, Tree::Node(Color::Red, _, _, _, _))
}

fn is_black_color(color: Color) -> bool {
    matches!(color, Color::Black)
}

fn balance_left(l: Tree, k: i64, v: bool, r: Tree) -> Tree {
    match l {
        Tree::Leaf => Tree::Leaf,
        Tree::Node(_, left, ky, vy, ry) => match (*left, *ry) {
            (Tree::Node(Color::Red, lx, kx, vx, rx), ryv) => Tree::Node(
                Color::Red,
                Box::new(Tree::Node(Color::Black, lx, kx, vx, rx)),
                ky,
                vy,
                Box::new(Tree::Node(Color::Black, Box::new(ryv), k, v, Box::new(r))),
            ),
            (ly, Tree::Node(Color::Red, lx, kx, vx, rx)) => Tree::Node(
                Color::Red,
                Box::new(Tree::Node(Color::Black, Box::new(ly), ky, vy, lx)),
                kx,
                vx,
                Box::new(Tree::Node(Color::Black, rx, k, v, Box::new(r))),
            ),
            (lx, rx) => Tree::Node(
                Color::Black,
                Box::new(Tree::Node(Color::Red, Box::new(lx), ky, vy, Box::new(rx))),
                k,
                v,
                Box::new(r),
            ),
        },
    }
}

fn balance_right(l: Tree, k: i64, v: bool, r: Tree) -> Tree {
    match r {
        Tree::Leaf => Tree::Leaf,
        Tree::Node(_, left, kx, vx, right) => match (*left, *right) {
            (Tree::Node(Color::Red, lx, ky, vy, rx), ry) => Tree::Node(
                Color::Red,
                Box::new(Tree::Node(Color::Black, Box::new(l), k, v, lx)),
                ky,
                vy,
                Box::new(Tree::Node(Color::Black, rx, kx, vx, Box::new(ry))),
            ),
            (lx, Tree::Node(Color::Red, ly, ky, vy, ry)) => Tree::Node(
                Color::Red,
                Box::new(Tree::Node(Color::Black, Box::new(l), k, v, Box::new(lx))),
                kx,
                vx,
                Box::new(Tree::Node(Color::Black, ly, ky, vy, ry)),
            ),
            (lx, rx) => Tree::Node(
                Color::Black,
                Box::new(l),
                k,
                v,
                Box::new(Tree::Node(Color::Red, Box::new(lx), kx, vx, Box::new(rx))),
            ),
        },
    }
}

fn ins(tree: Tree, kx: i64, vx: bool) -> Tree {
    match tree {
        Tree::Leaf => Tree::Node(Color::Red, Box::new(Tree::Leaf), kx, vx, Box::new(Tree::Leaf)),
        Tree::Node(Color::Red, a, ky, vy, b) => {
            if kx < ky {
                Tree::Node(Color::Red, Box::new(ins(*a, kx, vx)), ky, vy, b)
            } else if ky < kx {
                Tree::Node(Color::Red, a, ky, vy, Box::new(ins(*b, kx, vx)))
            } else {
                Tree::Node(Color::Red, a, ky, vy, Box::new(ins(*b, kx, vx)))
            }
        }
        Tree::Node(Color::Black, a, ky, vy, b) => {
            if kx < ky {
                if is_red(&a) {
                    balance_left(ins(*a, kx, vx), ky, vy, *b)
                } else {
                    Tree::Node(Color::Black, Box::new(ins(*a, kx, vx)), ky, vy, b)
                }
            } else if ky < kx {
                if is_red(&b) {
                    balance_right(*a, ky, vy, ins(*b, kx, vx))
                } else {
                    Tree::Node(Color::Black, a, ky, vy, Box::new(ins(*b, kx, vx)))
                }
            } else {
                Tree::Node(Color::Black, a, kx, vx, b)
            }
        }
    }
}

fn set_black(tree: Tree) -> Tree {
    match tree {
        Tree::Node(_, left, key, value, right) => Tree::Node(Color::Black, left, key, value, right),
        other => other,
    }
}

fn insert(tree: Tree, key: i64, value: bool) -> Tree {
    if is_red(&tree) {
        set_black(ins(tree, key, value))
    } else {
        ins(tree, key, value)
    }
}

fn set_red(tree: Tree) -> Tree {
    match tree {
        Tree::Node(_, left, key, value, right) => Tree::Node(Color::Red, left, key, value, right),
        other => other,
    }
}

fn make_black(tree: Tree) -> Del {
    match tree {
        Tree::Node(Color::Red, left, key, value, right) => Del(Tree::Node(Color::Black, left, key, value, right), false),
        other => Del(other, true),
    }
}

fn rebalance_left(c: Color, l: Tree, k: i64, v: bool, r: Tree) -> Del {
    match l {
        Tree::Node(Color::Black, _, _, _, _) => Del(balance_left(set_red(l), k, v, r), is_black_color(c)),
        Tree::Node(Color::Red, lx, kx, vx, rx) => Del(Tree::Node(Color::Black, lx, kx, vx, Box::new(balance_left(set_red(*rx), k, v, r))), false),
        _ => Del(Tree::Leaf, false),
    }
}

fn rebalance_right(c: Color, l: Tree, k: i64, v: bool, r: Tree) -> Del {
    match r {
        Tree::Node(Color::Black, _, _, _, _) => Del(balance_right(l, k, v, set_red(r)), is_black_color(c)),
        Tree::Node(Color::Red, lx, kx, vx, rx) => Del(Tree::Node(Color::Black, Box::new(balance_right(l, k, v, set_red(*lx))), kx, vx, rx), false),
        _ => Del(Tree::Leaf, false),
    }
}

fn del_min(tree: Tree) -> Delmin {
    match tree {
        Tree::Node(Color::Black, left, key, value, right) => match *left {
            Tree::Leaf => match *right {
                Tree::Leaf => Delmin(Del(Tree::Leaf, true), key, value),
                right_tree => Delmin(Del(set_black(right_tree), false), key, value),
            },
            left_tree => match del_min(left_tree) {
                Delmin(Del(lx, true), kx, vx) => Delmin(rebalance_right(Color::Black, lx, key, value, *right), kx, vx),
                Delmin(Del(lx, false), kx, vx) => Delmin(Del(Tree::Node(Color::Black, Box::new(lx), key, value, right), false), kx, vx),
            },
        },
        Tree::Node(Color::Red, left, key, value, right) => match *left {
            Tree::Leaf => Delmin(Del(*right, false), key, value),
            left_tree => match del_min(left_tree) {
                Delmin(Del(lx, true), kx, vx) => Delmin(rebalance_right(Color::Red, lx, key, value, *right), kx, vx),
                Delmin(Del(lx, false), kx, vx) => Delmin(Del(Tree::Node(Color::Red, Box::new(lx), key, value, right), false), kx, vx),
            },
        },
        Tree::Leaf => Delmin(Del(Tree::Leaf, false), 0, false),
    }
}

fn del(tree: Tree, key: i64) -> Del {
    match tree {
        Tree::Leaf => Del(Tree::Leaf, false),
        Tree::Node(color, left, kx, vx, right) => {
            if key < kx {
                match del(*left, key) {
                    Del(ly, true) => rebalance_right(color, ly, kx, vx, *right),
                    Del(ly, false) => Del(Tree::Node(color, Box::new(ly), kx, vx, right), false),
                }
            } else if key > kx {
                match del(*right, key) {
                    Del(ry, true) => rebalance_left(color, *left, kx, vx, ry),
                    Del(ry, false) => Del(Tree::Node(color, left, kx, vx, Box::new(ry)), false),
                }
            } else {
                match *right {
                    Tree::Leaf => {
                        if is_black_color(color) {
                            make_black(*left)
                        } else {
                            Del(*left, false)
                        }
                    }
                    right_tree => match del_min(right_tree) {
                        Delmin(Del(ry, true), ky, vy) => rebalance_left(color, *left, ky, vy, ry),
                        Delmin(Del(ry, false), ky, vy) => Del(Tree::Node(color, left, ky, vy, Box::new(ry)), false),
                    },
                }
            }
        }
    }
}

fn delete(tree: Tree, key: i64) -> Tree {
    match del(tree, key) {
        Del(next, _) => set_black(next),
    }
}

fn mk_map_aux(total: i64, n: i64, tree: Tree) -> Tree {
    if n == 0 {
        tree
    } else {
        let n1 = n - 1;
        let t1 = insert(tree, n1, n1 % 10 == 0);
        let t2 = if n1 % 4 == 0 { delete(t1, n1 + (total - n1) / 5) } else { t1 };
        mk_map_aux(total, n1, t2)
    }
}

fn main() {
    let handle = std::thread::Builder::new()
        .stack_size(64 * 1024 * 1024)
        .spawn(|| {
            println!("{}", count_true(&mk_map_aux(LIMIT, LIMIT, Tree::Leaf)));
        })
        .expect("spawn rbtree_del benchmark");
    handle.join().expect("run rbtree_del benchmark");
}

File: benchmarks/ocaml/rbtree_del.ml
type color =
| Red
| Black
;;

type tree =
| Leaf
| Node of color * tree * int * bool * tree
;;

type del = Del of tree * bool
;;

type delmin = Delmin of del * int * bool
;;

let limit = 100000
;;

let rec count_true tree =
  match tree with
  | Leaf -> 0
  | Node (_, left, _, value, right) ->
    count_true left + (if value then 1 else 0) + count_true right
;;

let is_red tree =
  match tree with
  | Node (Red, _, _, _, _) -> true
  | _ -> false
;;

let is_black_color color =
  match color with
  | Black -> true
  | _ -> false
;;

let balance_left l k v r =
  match l with
  | Leaf -> Leaf
  | Node (_, Node (Red, lx, kx, vx, rx), ky, vy, ry) ->
    Node (Red, Node (Black, lx, kx, vx, rx), ky, vy, Node (Black, ry, k, v, r))
  | Node (_, ly, ky, vy, Node (Red, lx, kx, vx, rx)) ->
    Node (Red, Node (Black, ly, ky, vy, lx), kx, vx, Node (Black, rx, k, v, r))
  | Node (_, lx, kx, vx, rx) ->
    Node (Black, Node (Red, lx, kx, vx, rx), k, v, r)
;;

let balance_right l k v r =
  match r with
  | Leaf -> Leaf
  | Node (_, Node (Red, lx, kx, vx, rx), ky, vy, ry) ->
    Node (Red, Node (Black, l, k, v, lx), kx, vx, Node (Black, rx, ky, vy, ry))
  | Node (_, lx, kx, vx, Node (Red, ly, ky, vy, ry)) ->
    Node (Red, Node (Black, l, k, v, lx), kx, vx, Node (Black, ly, ky, vy, ry))
  | Node (_, lx, kx, vx, rx) ->
    Node (Black, l, k, v, Node (Red, lx, kx, vx, rx))
;;

let rec ins tree kx vx =
  match tree with
  | Leaf -> Node (Red, Leaf, kx, vx, Leaf)
  | Node (Red, a, ky, vy, b) ->
    if kx < ky then Node (Red, ins a kx vx, ky, vy, b)
    else if ky < kx then Node (Red, a, ky, vy, ins b kx vx)
    else Node (Red, a, ky, vy, ins b kx vx)
  | Node (Black, a, ky, vy, b) ->
    if kx < ky then
      if is_red a then balance_left (ins a kx vx) ky vy b
      else Node (Black, ins a kx vx, ky, vy, b)
    else if ky < kx then
      if is_red b then balance_right a ky vy (ins b kx vx)
      else Node (Black, a, ky, vy, ins b kx vx)
    else Node (Black, a, kx, vx, b)
;;

let set_black tree =
  match tree with
  | Node (_, left, key, value, right) -> Node (Black, left, key, value, right)
  | other -> other
;;

let insert tree key value =
  if is_red tree then set_black (ins tree key value)
  else ins tree key value
;;

let set_red tree =
  match tree with
  | Node (_, left, key, value, right) -> Node (Red, left, key, value, right)
  | other -> other
;;

let make_black tree =
  match tree with
  | Node (Red, left, key, value, right) -> Del (Node (Black, left, key, value, right), false)
  | other -> Del (other, true)
;;

let rebalance_left c l k v r =
  match l with
  | Node (Black, _, _, _, _) -> Del (balance_left (set_red l) k v r, is_black_color c)
  | Node (Red, lx, kx, vx, rx) -> Del (Node (Black, lx, kx, vx, balance_left (set_red rx) k v r), false)
  | _ -> Del (Leaf, false)
;;

let rebalance_right c l k v r =
  match r with
  | Node (Black, _, _, _, _) -> Del (balance_right l k v (set_red r), is_black_color c)
  | Node (Red, lx, kx, vx, rx) -> Del (Node (Black, balance_right l k v (set_red lx), kx, vx, rx), false)
  | _ -> Del (Leaf, false)
;;

let rec del_min tree =
  match tree with
  | Node (Black, Leaf, key, value, right) ->
    (match right with
     | Leaf -> Delmin (Del (Leaf, true), key, value)
     | _ -> Delmin (Del (set_black right, false), key, value))
  | Node (Red, Leaf, key, value, right) ->
    Delmin (Del (right, false), key, value)
  | Node (color, left, key, value, right) ->
    (match del_min left with
     | Delmin (Del (lx, true), kx, vx) -> Delmin (rebalance_right color lx key value right, kx, vx)
     | Delmin (Del (lx, false), kx, vx) -> Delmin (Del (Node (color, lx, key, value, right), false), kx, vx))
  | Leaf -> Delmin (Del (Leaf, false), 0, false)
;;

let rec del tree key =
  match tree with
  | Leaf -> Del (Leaf, false)
  | Node (color, left, kx, vx, right) ->
    if key < kx then
      (match del left key with
       | Del (ly, true) -> rebalance_right color ly kx vx right
       | Del (ly, false) -> Del (Node (color, ly, kx, vx, right), false))
    else if key > kx then
      (match del right key with
       | Del (ry, true) -> rebalance_left color left kx vx ry
       | Del (ry, false) -> Del (Node (color, left, kx, vx, ry), false))
    else
      match right with
      | Leaf -> if is_black_color color then make_black left else Del (left, false)
      | _ ->
        (match del_min right with
         | Delmin (Del (ry, true), ky, vy) -> rebalance_left color left ky vy ry
         | Delmin (Del (ry, false), ky, vy) -> Del (Node (color, left, ky, vy, ry), false))
;;

let delete tree key =
  match del tree key with
  | Del (next, _) -> set_black next
;;

let rec mk_map_aux total n tree =
  if n = 0 then tree
  else
    let n1 = n - 1 in
    let t1 = insert tree n1 (n1 mod 10 = 0) in
    let t2 = if n1 mod 4 = 0 then delete t1 (n1 + (total - n1) / 5) else t1 in
    mk_map_aux total n1 t2
;;

let () =
  Printf.printf "%d\n" (count_true (mk_map_aux limit limit Leaf))
